vgscollect/src/main/java/com/verygoodsecurity/vgscollect/view/card/formatter/date/StrictDateFormatter.kt

```diff
@@ -42,130 +41,139 @@ internal open class StrictDateFormatter : BaseDateFormatter() {
-            indexStart in 0 until count -> runtimeData.substring(indexStart, indexEnd)
-            else -> ""
+        return if (splittedTime.size == 1) {
+            cacheMonth = ""
+            when {
+                isDeleteAction -> formattedYear
+                formattedYear.length < year!!.length -> formattedYear
+                else -> formattedYear + divider
+            }
+        } else {
+            val formattedMonth = formatMonth(splittedTime.getOrNull(1) ?: "")
+            if (formattedYear.isEmpty() && formattedMonth.isEmpty()) "" else formattedYear + divider + formattedMonth
         }
     }
 
-    private fun formatMonth(runtimeData: String): String {
-        val month = substringDateBlock(mounthIndex, mounthIndex+mounth!!.length, runtimeData)
+    private fun formatYear(year: String): String {
+        val isValid = patternYear?.matcher(year)?.matches() ?: true
 
-        if(month.isEmpty()) return runtimeData
+        val newM = when {
+            year.isEmpty() -> "".also { cacheYear = "" }
+            isValid -> year.also { cacheYear = it }
+            isDeleteAction -> "".also { moveCursorToEnd_Year() }
+            else -> cacheYear
+        }
 
-        val isValid = patternMounts.matcher(month).matches()
-        return if(isValid) {
-            return runtimeData
+        return newM
+    }
+
+    private fun moveCursorToEnd_Year() {
+        if (yearIndex > mounthIndex) {
+            source?.setSelection(runtimeData.length - 1)
         } else {
-            handleMonths(month, runtimeData)
+            source?.setSelection(0)
         }
     }
 
-    private fun handleMonths(month: String, runtimeData: String): String {
-        val formattedMonth = validateMonth(month)
-
-        val isValid = patternMounts.matcher(formattedMonth).matches()
-        return if(isValid) {
-            runtimeData.substring(0, runtimeData.length-1) + formattedMonth
+    private fun moveCursorToEnd_M() {
+        if (yearIndex < mounthIndex) {
+            source?.setSelection(runtimeData.length - 1)
         } else {
-            runtimeData.substring(0, runtimeData.length-1)
+            source?.setSelection(0)
         }
     }
 
-    private fun formatYear(runtimeData: String): String {
-        val year = substringDateBlock(yearIndex, yearIndex+year!!.length, runtimeData)
-
-        if(year.isEmpty()) return runtimeData
-
-        val isValid = patternYear?.matcher(year)?.matches()?:true
-        return if(isValid) {
-            return runtimeData
-        } else {
-            runtimeData.substring(0, runtimeData.length-1)
+    private fun formatMonth(month: String): String {
+        val isValid = patternMounts.matcher(month).matches()
+        val newM = when {
+            month.isEmpty() -> "".also { cacheMonth = "" }
+            isValid -> month.also { cacheMonth = it }
+            isDeleteAction -> "".also { moveCursorToEnd_M() }
+            !isDeleteAction -> validateMonth(month) ?: cacheMonth
```

Comment:
@Endorf Correct me if I'm wrong, ! isDeleteAction can be replaced with else, as we will always go inside this block of code if isDeleteAction == false?